src/dart/parse.rs

```diff
@@ -0,0 +1,1746 @@
+                kv.push((k, self.parse_expr()?));
+                if !self.eat_punctuation(',') {
+                    break;
+                }
+            }
+            self.expect_punctuation('}')?;
+            let kv_ty = if generics.is_empty() {
+                None
+            } else {
+                assert_eq!(generics.len(), 2);
+                let v_ty = generics.pop().unwrap();
+                let k_ty = generics.pop().unwrap();
+                Some((k_ty, v_ty))
+            };
+            return Ok(Box::new(Expr::Map { const_, kv_ty, kv }));
+        }
+        if const_ || self.eat_keyword("new") {
+            let ty = self.parse_type()?;
+            let ctor = if self.eat_punctuation('.') {
+                let name = self.parse_ident()?;
+                Some(name)
+            } else {
+                None
+            };
+            let args = self.parse_arguments()?;
+            return Ok(Box::new(Expr::New {
+                const_,
+                ty,
+                ctor,
+                args,
+            }));
+        }
+        if let Some(number) = self.try(|p| p.parse_number_literal()) {
+            return Ok(number);
+        }
+        if let Some(str_lit) = self.try(|p| p.parse_string_literal()) {
+            let mut strings = vec![str_lit];
+            while let Some(str_lit) = self.try(|p| p.parse_string_literal()) {
+                strings.push(str_lit);
+            }
+            return Ok(Box::new(Expr::String(strings)));
+        }
+        if self.eat_punctuation('#') {
+            return Ok(Box::new(Expr::Identifier(self.parse_ident()?)));
+        }
+        if let Ok(ident) = self.parse_ident() {
+            return Ok(Box::new(Expr::Identifier(ident)));
+        }
+        expected!(self, Expr);
+    }
+
+    fn parse_number_literal(&mut self) -> ParseResult<Box<Expr>> {
+        let mut number = String::new();
+
+        if let Some(Token::IntegerLiteral(int_part)) = self.cur {
+            self.bump_raw();
+            number += &int_part.as_str();
+        }
+
+        if self.cur == Some(Token::Punctuation('.')) {
+            self.bump_raw();
+            number.push('.');
+            if let Some(Token::IntegerLiteral(frac_part)) = self.cur {
+                self.bump_raw();
+                number += &frac_part.as_str();
+            }
+        }
+
+        if number.is_empty() {
+            expected!(self, NumberLiteral);
+        }
+
+        /*
+        if self.eat_keyword("e") || self.eat_keyword("E") {
+            if Some(Token::Punctuation('+')) == self.cur {
+                self.bump_raw();
+                number.push('+');
+            } else if Some(Token::Punctuation('-')) == self.cur {
+                self.bump_raw();
+                number.push('-');
+            }
+            if let Some(Token::IntegerLiteral(exp)) = self.cur {
+                self.bump_raw();
+                number += &exp.as_str();
+            }
+        }*/
```

Comment:
Remove this comment - the correct implementation will look a bit different and it's not necessary atm anyway.